Kaguya/Kaguya/Discord/Commands/Reference/Osu.cs

```diff
@@ -0,0 +1,121 @@
+using Kaguya.Database.Repositories;
+using Kaguya.Discord.DiscordExtensions;
+using Kaguya.External.Osu;
+using OsuSharp;
+
+namespace Kaguya.Discord.Commands.Reference
+{
+    [Module(CommandModule.Reference)]
+    [Group("osu")]
+    [Alias("o")]
+    public class Osu : KaguyaBase<Osu>
+    {
+        private readonly ILogger<Osu> _logger;
+        private readonly OsuClient _osuClient;
+        private readonly KaguyaUserRepository _kaguyaUserRepository;
+        private readonly KaguyaServerRepository _kaguyaServerRepository;
+        private const string GAME_MODE_STRING = "**Valid Game Modes:** `std`, `mania`, `taiko`, `ctb`";
+
+        public Osu(ILogger<Osu> logger, OsuClient osuClient, KaguyaUserRepository kaguyaUserRepository, 
+            KaguyaServerRepository kaguyaServerRepository) : base(logger)
+        {
+            _logger = logger;
+            _osuClient = osuClient;
+            _kaguyaUserRepository = kaguyaUserRepository;
+            _kaguyaServerRepository = kaguyaServerRepository;
+        }
+
+        [Priority(1)]
+        [Command("-set")]
+        [Summary("Give me an osu! username to remember you by! After setting a username, you will no longer have " +
+                 "to specify a username when using osu commands. You must also configure your game mode.\n\n" + GAME_MODE_STRING)]
+        [Remarks("<game mode> <username or id>")]
+        [Example("std name with spaces")]
+        [Example("mania 99999999")]
+        [Example("ctb really good ctb player")]
+        public async Task OsuSetCommand(string gameModeString, [Remainder] string username)
+        {
+            var data = new OsuData(username, gameModeString, _osuClient);
+            GameMode gameMode = data.GameMode;
+            
+            User osuUser = await data.GetOsuUserAsync();
+            if (osuUser == null)
+            {
+                await SendBasicErrorEmbedAsync(new OsuUserNotFoundException(username).Message);
+
+                return;
+            }
+            
+            KaguyaUser kaguyaUser = await _kaguyaUserRepository.GetOrCreateAsync(Context.User.Id);
+            kaguyaUser.OsuId = osuUser.UserId;
+            kaguyaUser.OsuGameMode = gameMode;
+
+            await _kaguyaUserRepository.UpdateAsync(kaguyaUser);
+            
+            await SendBasicEmbedAsync("Successfully set your osu! username: " + osuUser.Username.AsBold(), Color.Blue);
+        }
+
+        [Priority(2)]
+        [Command("-recent", RunMode = RunMode.Async)]
+        [Alias("-r")]
+        [Summary("Displays the most recent play for a given username. The `gamemode` parameter is only " +
+                 "optional if your username has been set via the `osu -set` command.\n\n" + GAME_MODE_STRING)]
+        [Remarks("\n<gamemode> [username]")]
+        // ReSharper disable once MethodOverloadWithOptionalParameter
+        public async Task OsuRecentCommand(string gameMode = null, [Remainder] string username = null)
```

Comment:
gameMode has no input validation, should this be validated?

Reply:
Validation is done in the OsuData.cs `GetGameMode()` function.